### PR TITLE
Trunk+abstracting injection flags

### PR DESCRIPTION
**Pull request by maximedenes**
This replaces #420 as discussed during the WG.

### CODE REVIEW

**Comment by maximedenes** [comment]
@herbelin would you mind rebasing this? Thanks!

**Comment by herbelin** [comment]
Rebased done. (It is actually more than a bugfix, it is the beginning of a proposal for adding a static control of tactic flags on top of the current dynamic control.)

**Comment by maximedenes** [comment]
@herbelin It seems there was a failure in `formal-topology`. Could you have a look?

**Comment by herbelin** [comment]
Overlay added for formal-topology (decide equality had become stronger).

File: tactics/equality.ml [hunk 1]
```diff
@@ -725,7 +733,14 @@ let _ =
...
 
-let find_positions env sigma ~no_discr t1 t2 =
+let keep_proof_equalities = function
+  | None -> !keep_proof_equalities_for_injection
+  | Some flags -> flags.keep_proof_equalities
+
+(* [keep_proofs] is relevant for types in Prop with elimination in Type *)
+(* In particular, it is relevant for injection but not for discriminate *)
+
+let find_positions keep_proofs env sigma ~no_discr t1 t2 =
```
Comment (maximedenes) [inline tactics/equality.ml]: `keep_proofs` should be a label here.

File: tactics/equality.ml [hunk 2]
```diff
@@ -772,20 +787,18 @@ let find_positions env sigma ~no_discr t1 t2 =
...
 
 let discriminable env sigma t1 t2 =
-  match find_positions env sigma ~no_discr:false t1 t2 with
+  match find_positions false env sigma ~no_discr:false t1 t2 with
     | Inl _ -> true
     | _ -> false
 
-let injectable env sigma t1 t2 =
-    match find_positions env sigma ~no_discr:true t1 t2 with
+let injectable keep_proofs env sigma t1 t2 =
```
Comment (maximedenes) [inline tactics/equality.ml]: Same here, should be a label.

File: plugins/funind/invfun.ml [hunk 3]
```diff
@@ -581,8 +586,8 @@ let rec reflexivity_with_destruct_cases g =
 		 | App(eq,[|_;t1;t2|]) when EConstr.eq_constr (project g) eq eq_ind ->
 		     if Equality.discriminable (pf_env g) (project g) t1 t2
 		     then Proofview.V82.of_tactic (Equality.discrHyp id) g
-		     else if Equality.injectable (pf_env g) (project g) t1 t2
-		     then tclTHENLIST [Proofview.V82.of_tactic (Equality.injHyp None id);thin [id];intros_with_rewrite]  g
+		     else if Equality.injectable false(*?*) (pf_env g) (project g) t1 t2
```
Comment (maximedenes) [inline plugins/funind/invfun.ml]: Is there something that shoud be checked, as indicated by the comment?

File: tactics/equality.ml [hunk 4]
```diff
@@ -1484,18 +1494,22 @@ let decompEqThen ntac (lbeq,_,(t,t1,t2) as u) clause =
...
   end
 
-let dEqThen with_evars ntac = function
-  | None -> onNegatedEquality with_evars (decompEqThen (ntac None))
-  | Some c -> onInductionArg (fun clear_flag -> onEquality with_evars (decompEqThen (ntac clear_flag))) c
+let use_keep_proofs = function
+  | None -> !keep_proof_equalities_for_injection
+  | Some b -> b
+
+let dEqThen keep_proofs with_evars ntac = function
```
Comment (maximedenes) [inline tactics/equality.ml]: label

Comment (maximedenes) [inline tactics/equality.ml]: label

File: tactics/equality.ml [hunk 5]
```diff
@@ -75,9 +83,9 @@ let _ =
 
 let injection_in_context = ref false
 
-let use_injection_in_context () =
-  !injection_in_context
-  && Flags.version_strictly_greater Flags.V8_5
+let use_injection_in_context = function
+  | None -> !injection_in_context && Flags.version_strictly_greater Flags.V8_5
```
Comment (maximedenes) [inline tactics/equality.ml]: What is this option and why do flags have priority over it? I'm not saying anything is wrong, just trying to understand.

**Comment by maximedenes** [comment]
@herbelin Is the topology fix backward compatible? If yes, can you make a PR upstream so that we can remove the overlay?

**Comment by herbelin** [comment]
It is not. We could make one of course but to the price of an uglier code or of getting rid of decide equality.

**Comment by maximedenes** [comment]
@herbelin Can you integrate the fix in topology?

**Comment by herbelin** [comment]
Sorry, I failed to understand the policy with fixes to contributor packages. We don't have control on `topology` besides overlays, do we? What do you want me to do exactly?

**Comment by maximedenes** [comment]
If you can make a pull request on topology, it would be ideal. But of course, if they don't ship a version of topology for each version of Coq, the fix should be backward compatible.

**Comment by herbelin** [comment]
@maximedenes: I don't know if you are a subscriber to `topology`. So, just to be sure you know, I made a PR, letting them decide what they think is best for them.

**Comment by maximedenes** [comment]
Indeed, I'm not, so thanks for the update.

**Comment by Zimmi48** [comment]
@herbelin if you give the link to the PR in the thread, it can help track its status.

**Comment by herbelin** [comment]
PR on topology [here](https://github.com/bmsherman/topology/pull/12).

**Comment by Zimmi48** [comment]
It seems that this PR creates warnings which make the Travis build fail (and in passing there is also a trivial merge conflict).

Maybe a mention should be added in the compatibility file (I'm not sure what's the rule concerning when to update this file).

Finally, following your PR on topology, a commit was added to the master branch for backward-forward compatibility. Maybe this branch can be tested as a temporary overlay to confirm this commit does the job?

**Comment by Zimmi48** [comment]
@herbelin Would you mind fixing the conflict and the warnings? Otherwise, this is going to miss 8.7+beta.

**Comment by bmsherman** [comment]
@Zimmi48 I have also now pushed that commit to the `ci` branch of the `bmsherman/topology` library, which is the one that I believe you are following. It took me a while because I wanted to be able to confirm that it was working on Travis, but have been having other difficulties with Travis so it wasn't building for some other reason.

**Comment by herbelin** [comment]
@Zimmi48: I rebased. I cannot reproduced the warning, so I suspect it resulted from a temporary failure of master.

**Comment by Zimmi48** [comment]
@herbelin Did you try to compile with the `-warn-error` configure option? Travis still reports some fatal warnings.

**Comment by Zimmi48** [comment]
Also note that you should normally be able to drop the overlay commit (2c1c901) now.

**Comment by herbelin** [comment]
@Zimmi48: Ah, sorry, I did not see the funind warning (I generally don't use the `-warn-error` configuration flag because I generally prefer to skip warnings while in work-in-progress phase; I should eventually find a more robust light way to check them before pushing though, other than my eyes jumping over the non-filtered `tacentries.ml` warning).

Warning now fixed, and overlay removed.

**Comment by Zimmi48** [comment]
> I generally don't use the -warn-error configuration flag because I generally prefer to skip warnings while in work-in-progress phase; I should eventually find a more robust light way to check them before pushing though

Same here.

**Comment by maximedenes** [comment]
A travis run on my repo seemed to indicate a failure in formal-topology, but maybe unrelated. A quick rebase would help figuring out.

**Comment by Zimmi48** [comment]
If we believe the Travis output before the conflict, the failure on formal-topology is real. This is also consistent with the history of this PR that has highlighted that formal-topology uses a lot of `decide equality`.
The conflicts come only from the `CHANGES` file and this is really annoying (this file is a large source of merge conflicts). We should probably come up with a strategy for generating the `CHANGES` file at the time of the release instead, but with formatted descriptions left in the commit messages (or PR messages). In the meantime, in the case of this PR, the `CHANGES` mention should go in the section "Changes beyond 8.7" (since it is too late at this point in the release cycle to backport a fix that affects the behavior of a tactic).

**Comment by bmsherman** [comment]
I'm sorry, for some reason I thought that I had pushed the change to stop using the `decide equality` tactic in the `formal-topology` library (`ci` branch of `bmsherman/topology`), but I just checked it now and apparently that change had not been incorporated. I've just made the change now.

(@Zimmi48 `formal-topology` actually only used `decide equality` once, and all that used to solve was decidable equality of the empty type anyway, so I removed the use of the `decide equality` tactic and just manually created the decidable equality instance.)

**Comment by herbelin** [comment]
It seems this branch has been forgotten because of a `needs:rebase` tag which was not relevant at that time and which noone checked the relevance about.

A property of such tag is of course that it eventually spontaneously becomes relevant again, so I rebased.

Anything preventing this fix to be merged? (I guess now in 8.8 since it changes the ML API?)

**Comment by Zimmi48** [comment]
> A property of such tag is of course that it eventually spontaneously becomes relevant again, so I rebased.

If you mean: "I rebased today", then you forgot to push. Not a big deal though because the only conflicts are in the change-log files.

**Pull request merged**